### PR TITLE
Fix JVM repository rules with Bzlmod

### PR DESCRIPTION
**Pull request by fmeum**
* Lets the `http_jar` and `jvm_{maven_}import_external` rules load from `@rules_java` as resolved within `@bazel_tools` rather than the calling module extension's repository.
* Uses the user-provided apparent repository name in `jvm_maven_import_external` to generate a target name for the jar that makes the shorthand `@my_import` form work with Bzlmod.
* Ensure that Bazel loads the `http_jar` rule with the full repo mapping for `bazel_tools` so that it sees `rules_java`, while still loading `http_archive` with a trivial repo mapping to prevent a cycle during module repo creation. This requires introducing a new key type for `BzlLoadFunction`.

### CODE REVIEW

**Comment by fmeum** [comment]
@bazel-io flag

**Comment by fmeum** [comment]
@meteorcloudy The darwin runner appears to be out of space again.

**Comment by keertk** [comment]
@bazel-io fork 7.0.0

**Comment by fmeum** [comment]
I pushed a new commit to realize https://github.com/bazelbuild/bazel/pull/20006#discussion_r1378054776.

**Review by Wyverald — APPROVED** [review]

**Comment by Wyverald** [comment]
to maybe save you some time, you need to add the two files under `tools/build_defs/repo/private` to bazel_fetch_test and bazel_overrides_test (https://cs.opensource.google/bazel/bazel/+/master:src/test/py/bazel/bzlmod/bazel_fetch_test.py;drc=12e17ed4eefa6e7eb9b3df44aded04ee4b5dadaa;l=64)

**Comment by fmeum** [comment]
> to maybe save you some time, you need to add the two files under `tools/build_defs/repo/private` to bazel_fetch_test and bazel_overrides_test (https://cs.opensource.google/bazel/bazel/+/master:src/test/py/bazel/bzlmod/bazel_fetch_test.py;drc=12e17ed4eefa6e7eb9b3df44aded04ee4b5dadaa;l=64)

Thanks, this does seem to fix `bazel_fetch_test`, but I don't see a similar setup in `bazel_overrides_test`. The failure there seems more concerning.

**Comment by fmeum** [comment]
I think that we need to prevent builtins injection for the new package while still using the non-trivial repo mapping. I will add another allowlist function.

**Comment by Wyverald** [comment]
ah, right, sorry about the oversight. having another allowlist does seem the best way forward.

**Comment by fmeum** [comment]
@Wyverald I pushed a fix. Both PRs should pass CI in isolation, but after one has been merged, I will need to modify the other.

**Review by Wyverald — APPROVED** [review]

**Comment by fmeum** [comment]
@Wyverald This approach doesn't seem to work since `labels.bzl` has to be loaded with a non-trivial repo mapping precisely if used for `http_jar`, but `http.bzl` has an eager load for `labels.bzl` that also triggers when loaded for `http_archive`. I don't see how we can solve this without a new key that allows us to separate these two situations.

**Comment by fmeum** [comment]
I reverted to the original approach with some additional cleanups (removing `isFileSafeForUninjectedEvaluation` in favor of the rule class match in `BlzmodRepoRuleCreator`. If you see another approach that doesn't require a new key type, I can of course switch to that.

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 7.0.0 RC5](https://releases.bazel.build/7.0.0/rc5/index.html). Please test out the release candidate and report any issues as soon as possible. Thanks!

**Comment by iancha1992** [comment]
The changes in this PR have been included in [Bazel 7.0.0 RC5](https://releases.bazel.build/7.0.0/rc5/index.html). Please test out the release candidate and report any issues as soon as possible. If you're using Bazelisk, you can point to the latest RC by setting USE_BAZEL_VERSION=last_rc.
Thanks!

File: tools/build_defs/repo/jvm.bzl [hunk 1]
```diff
@@ -136,7 +136,7 @@ def _jvm_import_external(repository_ctx):
         "",
         "alias(",
         "    name = \"%s\"," % extension,
-        "    actual = \"@%s\"," % repository_ctx.name,
+        "    actual = \"//:%s\"," % name,
```
Comment (gkdn) [inline tools/build_defs/repo/jvm.bzl:139]: I think this unintentionally changed the target of the alias.

For given
```
java_import_external(name = "foo", generated_rule_name = "bar")
```
The code used to create an alias for "foo". After this change, it generates an alias for "bar". 

As a result for java_import_external repositories that override default target by providing generated_rule_name (e.g. for annotation processors), the behavior for this target has changed.

Comment (fmeum) [reply tools/build_defs/repo/jvm.bzl:139]: Cc @Wyverald

Comment (gkdn) [reply tools/build_defs/repo/jvm.bzl:139]: https://github.com/bazelbuild/bazel/issues/24308

(FWIW, it was easy to workaround and reverting back the behavior could be also breaking change at this point if there is no other motivation to do so)

**Pull request closed without merging**